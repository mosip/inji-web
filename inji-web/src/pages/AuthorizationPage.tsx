import React, {useEffect, useState} from "react";
import {SpinningLoader} from "../components/Common/SpinningLoader";
import {api} from "../utils/api";
import {LandingPageWrapper} from "../components/Common/LandingPageWrapper";
import {ErrorSheildIcon} from "../components/Common/ErrorSheildIcon";
import {useTranslation} from "react-i18next";

export const AuthorizationPage: React.FC = () => {
    const url = window.location.href;
    const currentQueryParams = new URLSearchParams(window.location.search);
    const [error, setError] = useState<String>(currentQueryParams.get("error")+"");
    const {t} = useTranslation("AuthorizationPage");

    useEffect( () => {
        if(url.indexOf("error") === -1) {
            window.location.href = api.mimotoHost + "/authorize" + window.location.search;
        }
    },[])

    if(url.indexOf("error") == -1){
        return <div><LandingPageWrapper icon={<SpinningLoader/>} title={""} subTitle={""} gotoHome={false}/></div>
    }
    return <div><LandingPageWrapper icon={<ErrorSheildIcon />} title={t(`error.code.${error}`)} subTitle={t(`error.message.${error}`)} gotoHome={true}/></div>
}


