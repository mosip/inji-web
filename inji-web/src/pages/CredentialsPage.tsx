import React, {useEffect, useState} from "react";
import {useParams} from "react-router-dom";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {NavBar} from "../components/Common/NavBar";
import {CredentialList} from "../components/Credentials/CredentialList";
import {useDispatch, useSelector} from "react-redux";
import {storeSelectedIssuer} from "../redux/reducers/issuersReducer";
import {storeCredentials, storeFilteredCredentials} from "../redux/reducers/credentialsReducer";
import {api} from "../utils/api";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

import {ApiRequest, DisplayArrayObject, IssuerObject} from "../types/data";
import {getObjectForCurrentLanguage} from "../utils/i18n";
import {RootState} from "../types/redux";
import {isObjectEmpty} from "../utils/misc";

export const CredentialsPage: React.FC = () => {

    const {state, fetchRequest} = useFetch();
    const params = useParams<CredentialParamProps>();
    const dispatch = useDispatch();
    const {t} = useTranslation("CredentialsPage");
    const language = useSelector((state: RootState) => state.common.language);
    let displayObject= {} as DisplayArrayObject;
    let [selectedIssuer, setSelectedIssuer] = useState({} as IssuerObject)
    if(!isObjectEmpty(selectedIssuer)){
        displayObject = getObjectForCurrentLanguage(selectedIssuer.display, language);
    }

    useEffect(() => {
        const fetchCall = async () => {
            let apiRequest: ApiRequest = api.fetchSpecificIssuer;
            let response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeSelectedIssuer(response?.response));
            setSelectedIssuer(response?.response);

            apiRequest = api.fetchCredentialTypes2;
            response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            // apiRequest = api.fetchCredentialTypes;
            // response = await fetchRequest(
            //     apiRequest.url(response?.response[".well-known"]),
            //     apiRequest.methodType,
            //     apiRequest.headers()
            // );
            dispatch(storeFilteredCredentials(response?.response?.supportedCredentials));
            dispatch(storeCredentials(response?.response?.supportedCredentials));
        }
        fetchCall();
    }, [])

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }


    return <React.Fragment>
        <div className="bg-iw-background min-h-screen"
             data-testid="Credentials-Page-Container">
            <NavBar title={displayObject?.name} search={true} fetchRequest={fetchRequest} link={"/"}/>
            <div data-testid="Credential-List-Container" className="container mx-auto mt-8 px-10 sm:px-0">
                <CredentialList state={state}/>
            </div>
        </div>
    </React.Fragment>
}

