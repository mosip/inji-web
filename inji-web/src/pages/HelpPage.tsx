import React from "react";
import {HelpAccordion} from "../components/Help/HelpAccordion";
import {NavBar} from "../components/Common/NavBar";
import {useTranslation} from "react-i18next";

export const HelpPage: React.FC = () => {
    const {t} = useTranslation("HelpPage")
    return <React.Fragment>
        <div className={"bg-iw-background"} data-testid="Help-Page-Container">
            <NavBar title={t("title")} search={false} link={"/"}/>
            <div className="container mx-auto mt-8 ">
                <HelpAccordion/>
            </div>
        </div>
    </React.Fragment>
}
