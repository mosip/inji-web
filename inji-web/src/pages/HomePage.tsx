import React, {useEffect} from "react";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {IntroBox} from "../components/Common/IntroBox";
import {SearchIssuer} from "../components/Issuers/SearchIssuer";
import {IssuersList} from "../components/Issuers/IssuersList";
import {useDispatch} from "react-redux";
import {storeIssuers} from "../redux/reducers/issuersReducer";
import {api, MethodType} from "../utils/api";
import {IssuerObject} from "../types/data";
import {HeaderTile} from "../components/Common/HeaderTile";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

export const HomePage: React.FC = () => {

    const {state, fetchRequest} = useFetch();
    const dispatch = useDispatch();
    const {t} = useTranslation("HomePage");

    useEffect(() => {
        const fetchCall = async () => {
            const response = await fetchRequest(api.fetchIssuers(), MethodType.GET, null);
            const issuers = response?.response?.issuers.filter((issuer: IssuerObject) => issuer.credential_issuer === "Sunbird")
            dispatch(storeIssuers(issuers));
        }
        fetchCall();
    }, [])

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }

    return <div data-testid="Home-Page-Container">
        <div className="container mx-auto mt-8 px-4 flex-1 flex flex-col">
            <IntroBox/>
            <SearchIssuer state={state} fetchRequest={fetchRequest}/>
        </div>
        <HeaderTile content={t("containerHeading")}/>
        <IssuersList state={state}/>
    </div>

}
