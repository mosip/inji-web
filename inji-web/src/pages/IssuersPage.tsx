import React, {useEffect} from "react";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {IntroBox} from "../components/Common/IntroBox";
import {SearchIssuer} from "../components/Issuers/SearchIssuer";
import {IssuersList} from "../components/Issuers/IssuersList";
import {useDispatch} from "react-redux";
import {
    storeFilteredIssuers,
    storeIssuers
} from "../redux/reducers/issuersReducer";
import {api} from "../utils/api";
import {ApiRequest, IssuerObject} from "../types/data";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

export const IssuersPage: React.FC<IssuerPageProps> = ({className}) => {
    const {state, fetchRequest} = useFetch();
    const dispatch = useDispatch();
    const {t} = useTranslation("IssuersPage");

    useEffect(() => {
        const fetchCall = async () => {
            const apiRequest: ApiRequest = api.fetchIssuers;
            const response = await fetchRequest(
                apiRequest.url(),
                apiRequest.methodType,
                apiRequest.headers()
            );
            const issuers = response?.response?.issuers.filter(
                (issuer: IssuerObject) => issuer.protocol !== "OTP"
            );
            dispatch(storeFilteredIssuers(issuers));
            dispatch(storeIssuers(issuers));
        };
        fetchCall();
    }, []);

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }
    return (
        <div data-testid="Home-Page-Container">
            <div className="container mx-auto mt-8 flex flex-col px-4 sm:px-6 md:px-10 lg:px-20">
                <div className={className}>
                    <IntroBox />
                    <SearchIssuer state={state} fetchRequest={fetchRequest} />
                </div>
                <IssuersList state={state} />
            </div>
        </div>
    );
};

type IssuerPageProps = {
    className?: string;
};
