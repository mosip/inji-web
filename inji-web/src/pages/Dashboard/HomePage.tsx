import React, {useState, useEffect} from "react";
import {useTranslation} from "react-i18next";
import {IssuersPage} from "../IssuersPage.tsx";
import {convertStringIntoPascalCase} from "./utils.ts";
import {useUser} from "../../hooks/useUser.tsx";

export const HomePage: React.FC = () => {
    const {t} = useTranslation("Dashboard");
    const [displayName, setDisplayName] = useState<string | undefined>(undefined);
    const {user} = useUser();
    const userDisplayName = user?.displayName;

    useEffect(() => {
        setDisplayName(userDisplayName);
    }, [userDisplayName]);

    return (
        <div className="px-4 sm:px-6 md:px-10 lg:px-20 max-w-screen-xl mx-auto my-10">
            <h1 className="text-2xl sm:text-3xl lg:text-4xl font-medium text-[#04051D] text-center mb-10">
                {`${t("Home.welcome")} ${convertStringIntoPascalCase(
                    displayName
                )}!`}
            </h1>
            <IssuersPage />
        </div>
    );
};
