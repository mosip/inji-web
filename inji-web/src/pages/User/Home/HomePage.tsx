import React, {useState, useEffect} from 'react';
import {useTranslation} from 'react-i18next';
import {IssuersPage} from '../../IssuersPage';
import {useUser} from '../../../hooks/useUser';
import {convertStringIntoPascalCase} from "../../../utils/misc";
import {HomePageStyles} from "./HomePageStyles";

export const HomePage: React.FC = () => {
    const {t} = useTranslation('Dashboard');
    const [displayName, setDisplayName] = useState<string | undefined>(undefined);
    const {user} = useUser();
    const userDisplayName = user?.displayName;

    useEffect(() => {
        setDisplayName(userDisplayName);
    }, [userDisplayName]);

    return (
        <div className={HomePageStyles.container}>
            <h1 className={HomePageStyles.welcomeText}>
                {`${t('Home.welcome')} ${convertStringIntoPascalCase(
                    displayName
                )}!`}
            </h1>
            <IssuersPage />
        </div>
    );
};
