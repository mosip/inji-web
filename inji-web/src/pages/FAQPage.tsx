import React from 'react';
import {FAQAccordion} from '../components/Faq/FAQAccordion';
import {useTranslation} from 'react-i18next';
import {useLocation, useNavigate} from 'react-router-dom';
import {FAQPageProps} from '../components/Dashboard/types';
import {navigateToDashboardHome} from './Dashboard/utils';
import {PageTitle} from "../components/Common/PageTitle/PageTitle";
import { TertiaryButton } from './Dashboard/TertiaryButton';

export const FAQPage: React.FC<FAQPageProps> = ({backUrl, withHome}) => {
    const {t} = useTranslation(['FAQPage', 'Dashboard']);
    const navigate = useNavigate();
    const location = useLocation();
    const previousPagePath = location.state?.from;

    const handleBackClick = () => {
        if (backUrl) {
            navigate(backUrl); // Navigate to the URL sent by the parent
        } else if (previousPagePath) {
            navigate(previousPagePath); // Navigate to the previous link in history
        } else {
            navigateToDashboardHome(navigate); // Navigate to homepage if opened directly
        }
    };

    return (
        <div className="w-full flex flex-col sm:flex-row justify-between items-start p-1 sm:p-2 md:p-4 flex-grow overflow-y-auto">
            <div className="flex flex-row items-start gap-4 w-full">
                <div className="flex items-start">
                    <svg
                        width="29"
                        height="29"
                        viewBox="0 0 24 18"
                        fill="none"
                        xmlns="http://www.w3.org/2000/svg"
                        className="mr-2 cursor-pointer"
                        onClick={handleBackClick}
                    >
                        <path
                            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                            fill="#000000"
                        />
                    </svg>
                </div>
                <div className="flex flex-col items-start gap-8 w-full mr-2 sm:mr-4 md:mr-8 lg:mr-10">
                    <div className="flex flex-col items-start">
                        <PageTitle value={t('title')} testId={"faq"} />
                        {withHome && (
                            <TertiaryButton testId={'FAQ-Home-Button'} onClick={()=>navigateToDashboardHome(navigate)} title={t('Dashboard:Home.title')}/>
                        )}
                    </div>
                    <div className="w-full">
                        <FAQAccordion />
                    </div>
                </div>
            </div>
        </div>
    );
};
