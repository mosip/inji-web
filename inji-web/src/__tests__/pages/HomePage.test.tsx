import React, { useTransition } from 'react';
import { render, screen, fireEvent, waitFor,act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { HomePage } from '../../pages/HomePage';
import { toast} from 'react-toastify';

// Mock the toast function
jest.mock('react-toastify', () => {
  return {
    toast: {
      warning: jest.fn(),
    },
    ToastContainer: () => <div data-testid="toast-wrapper" />
  };
});

// Mock the components used in HomePage
jest.mock('../../components/Home/HomeBanner.tsx', () => ({
  HomeBanner: ({ onClick }: { onClick: () => void }) => (
    <div data-testid="HomeBanner" onClick={onClick}>HomeBanner</div>
  ),
}));
 
jest.mock('../../components/Home/HomeFeatures', () => ({
  HomeFeatures: () => <div data-testid="HomeFeatures">HomeFeatures</div>,
}));
 
jest.mock('../../components/Home/HomeQuickTip', () => ({
  HomeQuickTip: ({ onClick }: { onClick: () => void }) => (
    <div data-testid="HomeQuickTip" onClick={onClick}>HomeQuickTip</div>
  ),
}));
 
// Mock react-i18next
jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key,
  }),
}));


describe('HomePage', () => {
  const renderComponent = () => {
    return render(
      <BrowserRouter>
          <HomePage />
      </BrowserRouter>
    );
  };

  beforeEach(() => {
    // Clear mock before each test
    jest.clearAllMocks(); 
  });

  test('renders HomeBanner, HomeFeatures, and HomeQuickTip components', () => {
    renderComponent();
 
    expect(screen.getByTestId('HomeBanner')).toBeInTheDocument();
    expect(screen.getByTestId('HomeFeatures')).toBeInTheDocument();
    expect(screen.getByTestId('HomeQuickTip')).toBeInTheDocument();
  });
 
  test('navigates to /issuers when HomeBanner is clicked', () => {
    renderComponent();
 
    const homeBanner = screen.getByTestId('HomeBanner');
    fireEvent.click(homeBanner);
 
    expect(window.location.pathname).toBe('/issuers');
  });
 
  test('navigates to /issuers when HomeQuickTip is clicked', () => {
    renderComponent();
 
    const homeQuickTip = screen.getByTestId('HomeQuickTip');
    fireEvent.click(homeQuickTip);
 
    expect(window.location.pathname).toBe('/issuers');
  });
  
  test('shows toast only once when HomeQuickTip is clicked multiple times', async() => {
    renderComponent();
    const homeQuickTip = screen.getByTestId('HomeQuickTip');

    // Click HomeQuickTip multiple times
    fireEvent.click(homeQuickTip);
    fireEvent.click(homeQuickTip);
    fireEvent.click(homeQuickTip);

    // Check that toast.warning was called only once
    expect(toast.warning).toHaveBeenCalledTimes(1);
    expect(toast.warning).toHaveBeenCalledWith(
      'QuickTip.toastText', // String as the message
      expect.objectContaining({
        onClose: expect.any(Function),
      })
    );

    // Assertion to check if toast is rendered or not using test-id
    await waitFor(() => {
      const toastElement = screen.getByTestId("toast-wrapper");
      expect(toastElement).toBeInTheDocument();
    });
  });
  
  // Snapshot test cases
  test('matches snapshot for HomePage', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <HomePage />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  
  test('matches snapshot for HomePage with toast visible', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <HomePage />
      </BrowserRouter>
    );
    // Simulate toast visibility
    fireEvent.click(screen.getByTestId('HomeQuickTip'));
    expect(asFragment()).toMatchSnapshot();
  });

  test('toast disappears after timeout', async () => {

    // Create a more dynamic mock for toast
    jest.spyOn(toast, 'warning').mockImplementation((message, options) => {
      // Simulate toast lifecycle
      if (options && options.onClose) {
        setTimeout(options.onClose, 5000);
      }
      return 'mock-toast-id';
    });
  
    renderComponent();
    const homeQuickTip = screen.getByTestId('HomeQuickTip');
  
    // Click to show toast
    fireEvent.click(homeQuickTip);
  
    // Verify toast is initially present
    const initialToastWrapper = screen.getByTestId('toast-wrapper');
    expect(initialToastWrapper).toBeInTheDocument();
  
    // Wait for toast to potentially disappear
    await waitFor(() => {
      // Verify onClose was called
      expect(toast.warning).toHaveBeenCalledWith(
        'QuickTip.toastText',
        expect.objectContaining({
          onClose: expect.any(Function)
        })
      );
    }, { timeout: 6000 });
  });

  test('shows toast message again after it disappears and user clicks again', async () => {
    jest.useFakeTimers();

    // Create a mock for toast that triggers onClose after a short delay
    let onCloseCallback: (<T = unknown>(props?: T) => void) | undefined;
    jest.spyOn(toast, 'warning').mockImplementation((message, options) => {
      if (options && options.onClose) {
        onCloseCallback = options.onClose;
        setTimeout(() => {
          if (onCloseCallback) {
            onCloseCallback();
          }
        }, 100);
      }
      return 'mock-toast-id';
    });

    renderComponent();
    const homeQuickTip = screen.getByTestId('HomeQuickTip');

    // First click - show toast
    fireEvent.click(homeQuickTip);
    
    expect(toast.warning).toHaveBeenCalledTimes(1);

    // Wait for the toast to disappear 
    await act(async () => {
      jest.advanceTimersByTime(150);
    });

    // Reset the mock to track new calls
    jest.clearAllMocks();

    // Second click - should show toast again
    fireEvent.click(homeQuickTip);

    // Verify toast is shown again
    await waitFor(() => {
      expect(toast.warning).toHaveBeenCalledTimes(1);
    });

    await waitFor(() => {
      expect(toast.warning).toHaveBeenCalledWith(
        'QuickTip.toastText',
        expect.objectContaining({
          onClose: expect.any(Function)
        })
      );
    });

    // Cleanup timers
    jest.useRealTimers();
  });
});
