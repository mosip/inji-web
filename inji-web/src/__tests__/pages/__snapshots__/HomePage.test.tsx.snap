// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot for HomePage 1`] = `
<DocumentFragment>
  <div
    class="pb-20 flex flex-col gap-y-4 "
  >
    <div
      data-testid="HomeBanner"
    >
      HomeBanner
    </div>
    <div
      data-testid="HomeFeatures"
    >
      HomeFeatures
    </div>
    <div
      data-testid="HomeQuickTip"
    >
      HomeQuickTip
    </div>
    <div
      data-testid="toast-wrapper"
    />
  </div>
</DocumentFragment>
`;

exports[`HomePage matches snapshot for HomePage with toast visible 1`] = `
<DocumentFragment>
  <div
    class="pb-20 flex flex-col gap-y-4 "
  >
    <div
      data-testid="HomeBanner"
    >
      HomeBanner
    </div>
    <div
      data-testid="HomeFeatures"
    >
      HomeFeatures
    </div>
    <div
      data-testid="HomeQuickTip"
    >
      HomeQuickTip
    </div>
    <div
      data-testid="toast-wrapper"
    />
  </div>
</DocumentFragment>
`;
